tests/network/samba/samba_adcli.pm

```diff
@@ -0,0 +1,171 @@
+    assert_script_run("dig srv _ldap._tcp.$AD_hostname", fail_message => "failed to resolved the required LDAP SRV entry for AD controller");
+}
+
+sub join_domain {
+    # Join the Active Directory via `kinit`
+    script_retry("expect kinit.exp", retry => 3, timeout => 120, die => 1);
+    validate_script_output("klist", qr/$AD_domain/, fail_message => "Kerberos ticket for domain not listed");
+
+    # Retrying the adcli join is needed, due to https://bugs.freedesktop.org/show_bug.cgi?id=55487
+    # Joining the domain can take some time.
+    script_retry("adcli join -v --no-password --domain $AD_domain -U Administrator -C", delay => 30, retry => 3, timeout => 300, fail_message => "Joining AD domain failed (poo#96983)");
+    record_info("adcli info", script_output("adcli info -D '$AD_domain' -S '$AD_hostname' -v"));
+
+    #test samba with AD
+    # the wait_serial possibly could enter into a race condition, however for now this solution is good enough
+    # if something is not working in the future: i.e authentication is not working, switching to using expect
+    # would be a better idea
+    # TODO: REMOVED: -S '$AD_hostname'
+    assert_script_run("echo \"\$AD_DOMAIN_PASSWORD\" | net ads join --domain '$AD_domain' -U Administrator --no-dns-updates -i", timeout => 60, fail_message => "Error joining domain (poo#96986)");
+
+    #Enable pam authentication
+    assert_script_run "pam-config -a --mkhomedir";
+    assert_script_run "pam-config -a --sss";
+
+    foreach my $service (qw(smb nmb winbind sssd)) {
+        systemctl("enable --now $service");
+    }
+    systemctl('restart nscd');
+}
+
+sub update_password {
+    # Invalidate the password of the local computer account on AD
+    script_retry("adcli update --verbose --computer-password-lifetime=0 --domain '$AD_domain'", retry => 3, delay => 60, fail_message => "Error invalidating local password");
+    # Restore the password with --add-samba-data as requested by poo#91950
+    script_retry("adcli update --verbose --computer-password-lifetime=0 --domain '$AD_domain' --add-samba-data", retry => 3, delay => 60, fail_message => "Error re-adding password with samba data");
+
+    # Check the trust secret for the domain
+    if (script_run("wbinfo -tP") != 0) {
+        my $output = script_output('wbinfo -tP', proceed_on_failure => 1);
+
+        # Check for bsc#1188575
+        if ($output =~ "WBC_ERR_AUTH_ERROR") {
+            die("wbinfo output failed") unless (is_sle('=12-SP3') || is_sle('=12-SP4'));
+            record_soft_failure("bsc#1188575");
+        }
+    }
+}
+
+sub randomize_hostname {
+    my $hostname = "openqa-" . random_string(length => 8);
+    assert_script_run("hostnamectl set-hostname '$hostname'");
+}
+
+sub run {
+    select_serial_terminal;
+
+    # Ensure the required variables are set
+    get_required_var("AD_HOSTNAME");
```

Comment:
Why don't you do it on line 20 ?

Reply:
Because of `get_required_var`.

In line 20 we assign the var for the whole test run. If we would use `get_required_var` there, then I'm afraid that other test runs that don't hold this variable would fail. That's why I do `get_var` in line 20 and here `get_required_var` to ensure, that the variable is present, when this test run is executed.

Reply:
What other tests ? e.g. AD_HOSTNAME is all over the place, what other  test can use just one of the variable, that does not make sense to me or is just mess.

Reply:
Will check if that works.

Reply:
Yep that works. I moved all required vars up.